File: src/field/graphics/util/KeyEventMapping.java
package field.graphics.util;

import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;

import java.awt.event.KeyEvent;
import java.lang.reflect.Field;
import java.util.*;

import static org.lwjgl.glfw.GLFW.*;

public class KeyEventMapping {

	static HashMap<Integer, String> lookupMap = null;
	BiMap<Integer, Integer> keyCodes = HashBiMap.create();
	Set<Integer> glfwModifiers = new LinkedHashSet<>();
	Map<Integer, Character> forceTyped = new LinkedHashMap<>();


	public KeyEventMapping() {
		keyCodes.put(KeyEvent.VK_ENTER, GLFW_KEY_ENTER);
		keyCodes.put(KeyEvent.VK_UP, GLFW_KEY_UP);
		keyCodes.put(KeyEvent.VK_DOWN, GLFW_KEY_DOWN);
		keyCodes.put(KeyEvent.VK_LEFT, GLFW_KEY_LEFT);
		keyCodes.put(KeyEvent.VK_RIGHT, GLFW_KEY_RIGHT);

		keyCodes.put(KeyEvent.VK_PAGE_UP, GLFW_KEY_PAGE_UP);
		keyCodes.put(KeyEvent.VK_PAGE_DOWN, GLFW_KEY_PAGE_DOWN);


		keyCodes.put(KeyEvent.VK_ESCAPE, GLFW_KEY_ESCAPE);
		keyCodes.put(KeyEvent.VK_END, GLFW_KEY_END);
		keyCodes.put(KeyEvent.VK_HOME, GLFW_KEY_HOME);
		keyCodes.put(KeyEvent.VK_BACK_SPACE, GLFW_KEY_BACKSPACE);
		keyCodes.put(KeyEvent.VK_DELETE, GLFW_KEY_DELETE);
		keyCodes.put(KeyEvent.VK_TAB, GLFW_KEY_TAB);


		keyCodes.put(KeyEvent.VK_CONTROL, GLFW_KEY_LEFT_CONTROL);
//		keyCodes.put(KeyEvent.VK_CONTROL, GLFW_KEY_RIGHT_CONTROL);
		keyCodes.put(KeyEvent.VK_SHIFT, GLFW_KEY_LEFT_SHIFT);
		keyCodes.put(KeyEvent.VK_ALT, GLFW_KEY_LEFT_ALT);
		keyCodes.put(KeyEvent.VK_META, GLFW_KEY_LEFT_SUPER);
//		keyCodes.put(KeyEvent.VK_QUOTE, GLFW_KEY_APOSTROPHE);
//		keyCodes.put(KeyEvent.VK_QUOTEDBL, GLFW_KEY_APOSTROPHE);


		keyCodes.put(KeyEvent.VK_0, GLFW_KEY_0);
		keyCodes.put(KeyEvent.VK_Z, GLFW_KEY_Z);
		keyCodes.put(KeyEvent.VK_C, GLFW_KEY_C);
		keyCodes.put(KeyEvent.VK_G, GLFW_KEY_G);
		keyCodes.put(KeyEvent.VK_A, GLFW_KEY_A);
		keyCodes.put(KeyEvent.VK_V, GLFW_KEY_V);
		keyCodes.put(KeyEvent.VK_X, GLFW_KEY_X);
		keyCodes.put(KeyEvent.VK_I, GLFW_KEY_I);

		keyCodes.put(KeyEvent.VK_1, GLFW_KEY_1);
		keyCodes.put(KeyEvent.VK_2, GLFW_KEY_2);
		keyCodes.put(KeyEvent.VK_3, GLFW_KEY_3);
		keyCodes.put(KeyEvent.VK_4, GLFW_KEY_4);
		keyCodes.put(KeyEvent.VK_5, GLFW_KEY_5);
		keyCodes.put(KeyEvent.VK_6, GLFW_KEY_6);
		keyCodes.put(KeyEvent.VK_7, GLFW_KEY_7);
		keyCodes.put(KeyEvent.VK_8, GLFW_KEY_8);
		keyCodes.put(KeyEvent.VK_9, GLFW_KEY_9);

		keyCodes.put(KeyEvent.VK_SPACE, GLFW_KEY_SPACE);
		keyCodes.put(KeyEvent.VK_PERIOD, GLFW_KEY_PERIOD);


		glfwModifiers.add(GLFW_KEY_LEFT_CONTROL);
		glfwModifiers.add(GLFW_KEY_RIGHT_CONTROL);
		glfwModifiers.add(GLFW_KEY_LEFT_SHIFT);
		glfwModifiers.add(GLFW_KEY_RIGHT_SHIFT);
		glfwModifiers.add(GLFW_KEY_LEFT_ALT);
		glfwModifiers.add(GLFW_KEY_RIGHT_ALT);
		glfwModifiers.add(GLFW_KEY_LEFT_SUPER);
		glfwModifiers.add(GLFW_KEY_RIGHT_SUPER);

	}

	static public String lookup(int num) {

		if (lookupMap == null) {

			lookupMap = new HashMap<>();
			Field[] ff = org.lwjgl.glfw.GLFW.class.getFields();
			for (Field fff : ff) {
				try {
					if (fff.getName()
					       .startsWith("GLFW_KEY")) lookupMap.put(((Number) fff.get(null)).intValue(), fff.getName());
				} catch (IllegalAccessException e) {
				} catch (ClassCastException e) {

				}
			}
		}
		return lookupMap.get(num);
	}

	public Integer translateCode(int glfwcode) {
		return keyCodes.inverse()
			       .get(glfwcode);
	}

	public boolean isModifier(int glfwcode) {
		return glfwModifiers.contains(glfwcode);
	}

	public Character isForcedTyped(int glfwcode) {
		return forceTyped.get(glfwcode);
	}

}

File: src/main/java/field/graphics/util/KeyEventMapping.kt
package field.graphics.util

import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import org.lwjgl.glfw.GLFW
import java.awt.event.KeyEvent

//import org.cef.browser.CefBrowser_N;
class KeyEventMapping {
    var keyCodes: BiMap<Int, Int> = HashBiMap.create()
    var glfwModifiers: MutableSet<Int> = LinkedHashSet()
    var forceTyped: Map<Int, Char> = LinkedHashMap()
    fun translateCode(glfwcode: Int): Int? {
        return keyCodes.inverse()[glfwcode]
    }

    fun isModifier(glfwcode: Int): Boolean {
        return glfwModifiers.contains(glfwcode)
    }

    fun isForcedTyped(glfwcode: Int): Char? {
        return forceTyped[glfwcode]
    }

    companion object {
        var lookupMap: HashMap<Int, String>? = null
        @JvmStatic
		fun lookup(num: Int): String? {
            if (lookupMap == null) {
                lookupMap = HashMap()
                val ff = GLFW::class.java.fields
                for (fff in ff) {
                    try {
                        if (fff.name
                                .startsWith("GLFW_KEY")
                        ) lookupMap!![(fff[null] as Number).toInt()] = fff.name
                    } catch (e: IllegalAccessException) {
                    } catch (e: ClassCastException) {
                    }
                }
            }
            return lookupMap!![num]
        }
    }

    init {
        keyCodes[KeyEvent.VK_ENTER] = GLFW.GLFW_KEY_ENTER
        keyCodes[KeyEvent.VK_UP] = GLFW.GLFW_KEY_UP
        keyCodes[KeyEvent.VK_DOWN] = GLFW.GLFW_KEY_DOWN
        keyCodes[KeyEvent.VK_LEFT] = GLFW.GLFW_KEY_LEFT
        keyCodes[KeyEvent.VK_RIGHT] = GLFW.GLFW_KEY_RIGHT
        keyCodes[KeyEvent.VK_PAGE_UP] = GLFW.GLFW_KEY_PAGE_UP
        keyCodes[KeyEvent.VK_PAGE_DOWN] = GLFW.GLFW_KEY_PAGE_DOWN
        keyCodes[KeyEvent.VK_ESCAPE] = GLFW.GLFW_KEY_ESCAPE
        keyCodes[KeyEvent.VK_END] = GLFW.GLFW_KEY_END
        keyCodes[KeyEvent.VK_HOME] = GLFW.GLFW_KEY_HOME
        keyCodes[KeyEvent.VK_BACK_SPACE] = GLFW.GLFW_KEY_BACKSPACE
        keyCodes[KeyEvent.VK_DELETE] = GLFW.GLFW_KEY_DELETE
        keyCodes[KeyEvent.VK_TAB] = GLFW.GLFW_KEY_TAB
        keyCodes[KeyEvent.VK_CONTROL] = GLFW.GLFW_KEY_LEFT_CONTROL
        //		keyCodes.put(KeyEvent.VK_CONTROL, GLFW_KEY_RIGHT_CONTROL);
        keyCodes[KeyEvent.VK_SHIFT] = GLFW.GLFW_KEY_LEFT_SHIFT
        keyCodes[KeyEvent.VK_ALT] = GLFW.GLFW_KEY_LEFT_ALT
        keyCodes[KeyEvent.VK_META] = GLFW.GLFW_KEY_LEFT_SUPER
        //		keyCodes.put(KeyEvent.VK_QUOTE, GLFW_KEY_APOSTROPHE);
//		keyCodes.put(KeyEvent.VK_QUOTEDBL, GLFW_KEY_APOSTROPHE);
        keyCodes[KeyEvent.VK_0] = GLFW.GLFW_KEY_0
        keyCodes[KeyEvent.VK_Z] = GLFW.GLFW_KEY_Z
        keyCodes[KeyEvent.VK_C] = GLFW.GLFW_KEY_C
        keyCodes[KeyEvent.VK_G] = GLFW.GLFW_KEY_G
        keyCodes[KeyEvent.VK_A] = GLFW.GLFW_KEY_A
        keyCodes[KeyEvent.VK_SLASH] = GLFW.GLFW_KEY_SLASH
        keyCodes[KeyEvent.VK_BACK_SLASH] = GLFW.GLFW_KEY_BACKSLASH
        keyCodes[KeyEvent.VK_V] = GLFW.GLFW_KEY_V
        keyCodes[KeyEvent.VK_X] = GLFW.GLFW_KEY_X
        keyCodes[KeyEvent.VK_I] = GLFW.GLFW_KEY_I
        keyCodes[KeyEvent.VK_1] = GLFW.GLFW_KEY_1
        keyCodes[KeyEvent.VK_2] = GLFW.GLFW_KEY_2
        keyCodes[KeyEvent.VK_3] = GLFW.GLFW_KEY_3
        keyCodes[KeyEvent.VK_4] = GLFW.GLFW_KEY_4
        keyCodes[KeyEvent.VK_5] = GLFW.GLFW_KEY_5
        keyCodes[KeyEvent.VK_6] = GLFW.GLFW_KEY_6
        keyCodes[KeyEvent.VK_7] = GLFW.GLFW_KEY_7
        keyCodes[KeyEvent.VK_8] = GLFW.GLFW_KEY_8
        keyCodes[KeyEvent.VK_9] = GLFW.GLFW_KEY_9
        keyCodes[KeyEvent.VK_SPACE] = GLFW.GLFW_KEY_SPACE
        keyCodes[KeyEvent.VK_PERIOD] = GLFW.GLFW_KEY_PERIOD
        glfwModifiers.add(GLFW.GLFW_KEY_LEFT_CONTROL)
        glfwModifiers.add(GLFW.GLFW_KEY_RIGHT_CONTROL)
        glfwModifiers.add(GLFW.GLFW_KEY_LEFT_SHIFT)
        glfwModifiers.add(GLFW.GLFW_KEY_RIGHT_SHIFT)
        glfwModifiers.add(GLFW.GLFW_KEY_LEFT_ALT)
        glfwModifiers.add(GLFW.GLFW_KEY_RIGHT_ALT)
        glfwModifiers.add(GLFW.GLFW_KEY_LEFT_SUPER)
        glfwModifiers.add(GLFW.GLFW_KEY_RIGHT_SUPER)

//		forceTyped.put(GLFW_KEY_DELETE, (char)0x7f);
    }
}